assets/translate.js

```diff
@@ -81,29 +83,30 @@ $( function () {
  */
 $( function () {
 	var sourceLangWidget,
+		model = new App.Model( appConfig.translations ),
 		$sourceLangWidget = $( '.source-lang-widget' );
-	if ( $sourceLangWidget.length !== 1 ) {
-		// Don't do anything if the widget isn't present.
-		return;
+
+	// Change the source lang value.
+	if ( $sourceLangWidget.length === 1 ) {
+		sourceLangWidget = OO.ui.infuse( $sourceLangWidget[ 0 ] );
+		sourceLangWidget.on( 'change', function () {
+			model.setSourceLang( sourceLangWidget.getValue() );
+		} );
 	}
-	sourceLangWidget = OO.ui.infuse( $sourceLangWidget[ 0 ] );
-	sourceLangWidget.on( 'change', function () {
-		var newLangCode = sourceLangWidget.getValue();
-		// Go through all the field labels and fetch new values from the translations.
-		$( '.translation-fields .oo-ui-fieldLayout' ).each( function () {
-			var fieldLayout = OO.ui.infuse( $( this ) ),
-				nodeId = fieldLayout.getField().data[ 'tspan-id' ];
-			if ( appConfig.translations[ nodeId ][ newLangCode ] === undefined ) {
-				// If there's no source language available for a string,
-				// show a message and the fallback language.
-				fieldLayout.setLabel( $.i18n(
-					'source-lang-not-found',
-					[ appConfig.translations[ nodeId ].fallback.text ]
-				) );
-				fieldLayout.$element.addClass( 'source-lang-not-found' );
+
+	// Change field labels when the source lang changes.
+	$( '.translation-fields .oo-ui-fieldLayout' ).each( function () {
+		var fieldLayout = OO.ui.infuse( $( this ) );
+		model.on( 'sourceLangSet', function () {
+			var nodeId = fieldLayout.getField().data[ 'tspan-id' ],
+				sourceTranslation = model.getSourceTranslation( nodeId );
+			if ( sourceTranslation.class === 'source-lang-not-found' ) {
+				fieldLayout.setLabel(
+					$.i18n( 'source-lang-not-found', [ sourceTranslation.label ] )
```

Comment:
I see what you mean about using the i18n in the model. I think for simplicity, you can keep it here, but if you use it in the model, you can probably get away with not bothering with the whole condition here. Then again, this is probably more logical *and* saves us the trouble of adding jquery.i18n to the model, so let's not make things difficult.

... But you were right :D